### ISSUE TITLE
Charts: deferred renderer sometimes called multiple times, causing errors

### ISSUE THREAD

**Issue by metaxmx** [primefaces/primefaces#10839]
### Describe the bug

In the ChartJS components (LineChart, BarChart etc.) there is a problem, that when you refresh the component more than once, while it is invisible (e.g. in another TabView tab), there is an error during rendering, when the component is finally visible again.

E.g.:
```
Uncaught Error: Canvas is already in use. Chart with ID '1' must be destroyed before the canvas with ID 'frmTest:tabView:chart_canvas' can be reused.
    bn http://localhost:8080/primefaces-test/jakarta.faces.resource/chartjs/chartjs.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:13
    _render http://localhost:8080/primefaces-test/jakarta.faces.resource/chartjs/chartjs.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:96
    render http://localhost:8080/primefaces-test/jakarta.faces.resource/core.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:5810
    renderDeferred http://localhost:8080/primefaces-test/jakarta.faces.resource/core.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:5794
    invokeDeferredRenders http://localhost:8080/primefaces-test/jakarta.faces.resource/core.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:1413
    postTabShow http://localhost:8080/primefaces-test/jakarta.faces.resource/components.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:30309
    show http://localhost:8080/primefaces-test/jakarta.faces.resource/components.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:30092
    oncomplete http://localhost:8080/primefaces-test/jakarta.faces.resource/components.js.xhtml?ln=primefaces&v=14.0.0-SNAPSHOT:30137
```

I tried to deduce the root cause of this problem, and apparently the ChartJs components are implemented in JavaScript as `DeferredWidget`. In its current implementation, a `DeferredWidget` won't call a direct rendering when an invisible widget is initialized or refreshed, but instead add a callback to render, after the widget gets visible again.
However, if you refresh a `DeferredWidget` twice (or more), while it is invisible, the `DeferredWidget` will "stack" the deferred renderings, so when component get visible again, the `_render()` function will be called twice (or more often) in a row.
In case of the ChartJs widget this will cause an error, because the first `_render()` call initializes a ChartJs instance on the canvas, while the second `_render()` call can't do that as the canvas is already in use.

Generally, I'm not sure if a `DeferredWidget` should plan multiple deferred renderings while invisible anyways. In my opinion the `DeferredWidget` should detect whether there is already a deferred rendering planned, and then not plan another one.
But in case this is how you want the `DeferredWidget` to behave, then I think the rendering of the ChartJs widgets should be made idempotent, so when two (or more) `_render()` calls are executed in a row, the widget does not produce an error and still show a functioning ChartJs graph.

This problem exists in the latest master version (14.0.0-SNAPSHOT), as well as in the 13.x branch.

### Reproducer

I created a reproducer in the forked PrimeFaces-Test repository:
https://github.com/metaxmx/primefaces-test

- master-Branch: Commit [48f771add1e7430f201432e9635d5e5a4588bbe3](https://github.com/metaxmx/primefaces-test/tree/48f771add1e7430f201432e9635d5e5a4588bbe3)
- jakarta-Branch: Commit [d97a0eedb03f797bd2829d3a427c3e1e02d78ce4](https://github.com/metaxmx/primefaces-test/tree/d97a0eedb03f797bd2829d3a427c3e1e02d78ce4)

How to Reproduce:

1. Start Test-App and open test.xhtml
2. Open Developer Console
3. Switch Tab, so the Chart is invisible
4. Press the "Refresh Chart" button multiple times (which will update the chart per AJAX)
5. Switch back to Chart tab, so it is visible again
6. An error from ChartJS will be displayed on the javascript console

### Expected behavior

When the chart gets visibile again, it should initialize only one ChartJS instance, and not try to create multiple, overlapping charts

### PrimeFaces edition

Community

### PrimeFaces version

14.0.0-SNAPSHOT

### JSF implementation

Mojarra

### JSF version

4.0

### Java version

20

### Browser(s)

Firefox

**Comment by melloware** [comment]
Good debugging @metaxmx !   I am wondering if this method.

```js
addDeferredRender: function(widgetId, containerId, fn) {
     this.deferredRenders.push({widget: widgetId, container: containerId, callback: fn});
}
```

Shouldn't blindly push the widget onto the array but should remove an existing one with that `widgetId` first before add so it would be the "last" version of that widget pushed onto the stack?

**Comment by melloware** [comment]
OK PR submitted all integration tests passed.

**Comment by metaxmx** [comment]
Thank you @melloware for the quick response and fix! Now everything works fine

**Comment by melloware** [comment]
Thanks for debugging and testing!

**Comment by tandraschko** [comment]
@metaxmx @melloware 
this commit breaks a simple case:

```
<h:body>

    <p:dialog widgetVar="dlg">
        <p:accordionPanel activeIndex="" multiple="true" >
            <p:tab>
                <p:inputTextarea rows="3" style="width: 100%">
            
                </p:inputTextarea>
            </p:tab>
        </p:accordionPanel>
    </p:dialog>
    
    <button onclick="PF('dlg').show(); return false;">open dlg</button>
    
</h:body>
```

inputTextArea misses the _render JS call, therefore no handlers like focus handling etc is applied.
just click into the inputTextArea, borders are not colored.

please check this, otherwise we need to revert for now.

**Comment by melloware** [comment]
I will check it out.

**Comment by melloware** [comment]
@metaxmx can you try this MonkeyPatch and make sure your scenario is still working. Basically widgetID is not enough it has to be the widgetId and containerId combo before removing it.

```js
PrimeFaces.addDeferredRender = function(widgetId, containerId, fn) {
    // remove existing
    for (var i = (this.deferredRenders.length - 1); i >= 0; i--) {
        var deferredRender = this.deferredRenders[i];

        if (deferredRender.widget === widgetId && deferredRender.container === containerId) {
            this.deferredRenders.splice(i, 1);
        }
    }

    // add new
    this.deferredRenders.push({
        widget: widgetId,
        container: containerId,
        callback: fn
    });
};
```

**Comment by metaxmx** [comment]
Hi, my gut feeling would teel me, that we still might run into problem with this new implementation, if a DeferredWidget gets refreshed, while having multiple layers of invisible containers, where each container will trigger a `addDeferredRender ()` with their container id.

However, I tried to reproduce the bug with your changed code, and even after wrapping the Chart into multiple layers of tabs, and refreshing it while different tabviews were currently making the chart invisible, I could not reproduce the original problem and everything worked fine.

So probably your revised fix does still solve the bug, and I'm happy if we keep it in the new form, instead of reverting the bugfix alltogether.

**Comment by melloware** [comment]
Thanks for testing @metaxmx  .  @tandraschko you are safe to test it tomorrow your scenario above is now working for me.

**Comment by tandraschko** [comment]
works fine again, thanks

**Comment by 99sono** [comment]
I will very likely soon open an issue for PrimeFaces 6.2 on “deferredRenders.”

I have now spotted a JavaScript memory leak with the array:

```javascript
PrimeFaces.deferredRenders
```

growing very large on browser tabs that are idle but where push events keep on updating the screen over time.

A hidden popup with widgets, like PrimeFaces widgets, seems to keep leaking deferred renderers over time.

Then it takes opening the popup once to clear the memory leak.

I will hopefully create the new issue soon.

**Comment by tandraschko** [comment]
If you can still replicate it on 15.0.0-RC2, please do.
Otherwise 6.2 is end of life since many years

**Comment by melloware** [comment]
@99sono yes this was fixed years ago you should upgrade off of 6.2

**Comment by 99sono** [comment]
# Thanks For the feedback (upgrading primefaces is not without risk)
Here is the analysis I received. Essentially, it aligns with what both of you suggested: upgrading the library version to something more modern.

Upgrading core libraries like PrimeFaces is something we prefer to do in major new releases. It is very risky to upgrade a library as important as PrimeFaces without thorough consideration. This is a very core component and would require retesting the entire application. It is often safer to back-port a patch than just go for a major upgrade.

# LLM Feedback
In any case, here is the feedback from Claude:

Based on the provided code and context, I’ll analyze the memory leak issue:

1. Page Structure Analysis:
```html
- p:dialog (obsucatedName01RegionControlAndMaterialFlowControlEventsPopup)
  - h:form (eventDialogForm)
    - h:panelGroup (mainPopupContentGroup)
      - p:dataTable (eventsTable)
```
The key components are:
- A modal dialog
- A form containing the content
- A data table for events that's scrollable and selectable

2. DeferredRenders Accumulation:
The accumulated entries point to:
- Container: dialog ID
- Widget: data table ID 

This suggests the p:dataTable is being registered as a deferred widget multiple times during push events.

3. Technical Analysis:

a) Potential Issues in Our Code:
- The push updates are triggering full re-renders of components that contain deferred widgets
- The dialog being hidden doesn't properly cleanup its deferred renders

b) PrimeFaces Deferred Widgets in Use:
```javascript
- p:dataTable (implements DeferredWidget behavior)
- p:dialog (may affect child widget initialization)
```

c) PrimeFaces Implementation Issues in 6.2:
```javascript
addDeferredRender: function(widgetId, containerId, fn) {
    this.deferredRenders.push({widget: widgetId, container: containerId, callback: fn});
}
```
The code doesn't:
- Check for duplicate entries
- Clean up old entries for same widget/container
- Handle hidden container scenarios properly

4. Root Cause:
The memory leak occurs because:
1. Each push event triggers re-rendering
2. The data table registers as deferred widget
3. PrimeFaces 6.2 keeps adding entries without cleanup
4. Hidden dialogs complicate widget lifecycle

This was fixed in newer PrimeFaces versions by improving deferred widget management and cleanup.

Solution Options:
1. Upgrade PrimeFaces (recommended)
2. Minimize updates when dialog is hidden (current approach)
3. Consider manual cleanup of deferredRenders array

The current fix avoiding updates while hidden should help, but the core issue is in PrimeFaces 6.2's widget management.


# Prompt to obtain the above feedback

I have now copied the original, non-minified source code of `primefaces-core.js` into my project.

There are several files in the directory:
```
C:\dev\dump\MemoryLeak_EventsButton\primefaces_library\primefaces-6.2-sources\META-INF\resources\primefaces\core
```
but I believe the memory leaks are likely part of either of these two files: `core.js` and `core.widget.js`.

Please note, I have not yet verified if the changes to the 
```
obfuscatedName01.transport.web.touch.framework.mainpopup.obfuscatedName01TouchBoxMainPopupPushEventPageBeanProxy#getRefreshComponentIds
```
with the new IF logic in there make any difference.

But I would like your take on this memory leak.

Essentially, the problem is as follows:
Originally, the bug reported that our screen would freeze if it had been idle for a while and a person tried to press the "events" button to open the main screen popup.

I eventually found out that the reason for this was a memory leak in the array:
```
PrimeFaces.deferredRenders = []
```
As this array starts accumulating entries over time like this:
```
The array contains elements with:

container "obfuscatedName01RegionControlAndMaterialFlowControlEventsPopup"

widget "eventDialogForm:eventsTable"
```
Some also have:
```
container undefined

widget "eventDialogForm:eventsTable"
```
To me, this feels like a memory leak because each time one of these entries is added to the array, if other entries like it were already in there, I assume they should be eliminated and replaced by the new elements instead of just stacking them.

Yesterday, I was further able to reproduce the bug by simulating the creation of events and their subsequent clearing.

Each time the button would navigate between grey "OK" to RED "error," the leak would occur.

My suspicion is because we were all the time rendering the `BROWSER_CLIENT_ID_EVENTS_POPUP_CONTENT`.

I have yet to verify that I fixed this bug now with this code change.

But I would like you to consider the PrimeFaces 6.2 code I have attached, as well as the popup dialog XHTML page elements that I have attached.

Essentially, they are in a chain of includes with one another.

I do not know the details of the PrimeFaces widgets that I used, like `p:dataTable`, to say if they are "deferredWidgets" or not.

But I assume some of these must be.

One more thing that might be relevant to you:

The reason opening the popup becomes so excruciatingly slow is because when the array of the deferred renderers is packed with records in there, once you open up, it will spend a lot of time in the `invokeDeferredRenderers` logic where the callback logic relating to rendering widgets is super expensive.

Finally, the `invokeDeferredRenderers` has a healing effect as it removes the deferredRenderers out of the array.

But at that point, the price has been paid.

Can you please:
1. Understand the page structure of the PrimeFaces dialog with the attached XHTML.
2. Reason about the fact that the deferredRenderers array is accumulating elements with:
```
container "obfuscatedName01RegionControlAndMaterialFlowControlEventsPopup"

widget "eventDialogForm:eventsTable"
```
each time the push event is triggered to make events pending and cleared.
3. Study the code of PrimeFaces and make an informed opinion on:
   (a) Do we do something wrong in our code or page structure?
       Note: Sure, we can optimize to try to not re-render useless parts of the page while a popup is closed. But PrimeFaces should support this inefficient scenario anyway.
   (b) Summarize the page elements that we use that would likely qualify as being PrimeFaces widgets with support of deferred rendering behavior.
   (c) Determine if the logic of PrimeFaces around deferred renders is buggy. When adding a deferred renderer to the array, they do not take any preventive measures to mitigate the risk of a leak.

# Next Steps

First, I will determine if my local fix, which takes into account whether the popup is open or not and avoids re-rendering invisible popup content, is effective. Second, I will look at more recent versions of PrimeFaces to see what these JavaScript files might look like there. I will post feedback once I know more.

**Comment by 99sono** [comment]
I think i found the proper fix for the issue in the latest release of primefaces.
see here:
https://github.com/primefaces/primefaces/blob/15.0.0/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.js

```javascript
  /**
         * Some widgets need to compute their dimensions based on their parent element(s). This requires that such
         * widgets are not rendered until they have become visible. A widget may not be visible, for example, when it
         * is inside a tab that is not shown when the page is rendered. PrimeFaces provides a global mechanism for
         * widgets to render once they are visible. This is done by keeping a list of widgets that need to be rendered,
         * and checking on every change (AJAX request, tab change etc.) whether any of those have become visible. A
         * widgets should extend `PrimeFaces.widget.DeferredWidget` to make use of this functionality.
         *
         * Adds a deferred render to the global list.  If this widdget has already been added only the last instance
         * will be added to the stack.
         *
         * @param {string} widgetId The ID of a deferred widget.
         * @param {string} containerId ID of the container that should be visible before the widget can be rendered.
         * @param {() => boolean} fn Callback that is invoked when the widget _may_ possibly have become visible. Should
         * return `true` when the widget was rendered, or `false` when the widget still needs to be rendered later.
         */
        addDeferredRender: function(widgetId, containerId, fn) {
            // remove existing
            this.deferredRenders = this.deferredRenders.filter(deferredRender => {
                return !(deferredRender.widget === widgetId && deferredRender.container === containerId);
            });

            // add new
            this.deferredRenders.push({ widget: widgetId, container: containerId, callback: fn });
        },
```

The  // remove existing step is key.

I need to check if I can locally patch code with that logic.
NOTE: My local fix of not re-endering the popup content when it is not visible on the UI did do the magic, but I remain uncertain if will have other page sections leaking like this.

So I will be happy if the javascript above can actually work on 6.2.

I will report back soon.

**Comment by 99sono** [comment]
Seems to work fine the patch of the javascript.
So the fix i apply is two fold.
On the one hand, yes I optimize the logic that is deciding what elements of the page to refresh to take into account if the popup is open or not and if the popup is currently not open I no longer re-render the internal contents.
But this is a fix that uses for the specific leak i encountered but not applicable application wide.


On the other hand i got to the top level xhtml page template inherited by most other pages.
     <script type="text/javascript" src="#{resource['script:primefaces/TRK35672_addDeferredRender_memory_leak.js']}"/>

Put in there this little snippet.
And sneak into that javascript the following:

```
/**
 * PrimeFaces versions 6.2 and higher have a serious bug causing a memory leak
 * in the PrimeFaces.deferredRenders array.
 * See the bug report: hidden URL 
 *
 * The issue occurs when a page refreshes areas that are not currently visible
 * (e.g., refreshing the content of a hidden popup). PrimeFaces defers the rendering
 * of the widget's content, but it blindly adds new deferred rendering logic to an array
 * without checking for existing entries.
 *
 * This snippet of code is adapted from:
 * https://github.com/primefaces/primefaces/blob/15.0.0/primefaces/src/main/resources/META-INF/resources/primefaces/core/core.widget.js
 *
 * The crucial step here is the "remove existing" logic that was missing.
 *
 */
if (PrimeFaces && PrimeFaces.addDeferredRender) {
    PrimeFaces.addDeferredRender = function (widgetId, containerId, fn) {
        // Remove existing deferred renders with the same widgetId and containerId
        this.deferredRenders = this.deferredRenders.filter(deferredRender => {
            return !(deferredRender.widget === widgetId && deferredRender.container === containerId);
        });

        // Add new deferred render
        this.deferredRenders.push({widget: widgetId, container: containerId, callback: fn});
    };
}
```
